In [0]:
from pyspark.sql.functions import *

In [0]:
file_path = "/databricks-datasets/learning-spark-v2/sf-fire/sf-fire-calls.csv"

raw_sf_fire_calls_df = spark.read.format("csv")\
                            .option("inferSchema","true")\
                            .option("header","true")\
                            .load(file_path)

display(raw_sf_fire_calls_df)

Call Number Unit ID Incident Number CallType Call Date Watch Date Call Final Disposition Available DtTm Address City Zipcode of Incident Battalion Station Area Box OrigPriority Priority Final Priority ALS Unit Call Type Group NumAlarms UnitType Unit sequence in call dispatch Fire Prevention District Supervisor District Neighborhood Location RowID Delay 20110014 M29 2003234 Medical Incident 2002-01-11 2002-01-10 Other 01/11/2002 01:58:43 AM 10TH ST/MARKET ST SF 94103 B02 36 2338 1 1 2 true null 1 MEDIC 1 2 6 Tenderloin (37.7765408927183, -122.417501464907) 020110014-M29 5.233333333333333 20110015 M08 2003233 Medical Incident 2002-01-11 2002-01-10 Other 01/11/2002 02:10:17 AM 300 Block of 5TH ST SF 94107 B03 08 2243 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7792841462441, -122.402061300134) 020110015-M08 3.0833333333333335 20110016 B02 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 6 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B02 3.05 20110016 B04 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:51:54 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 3 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B04 2.316666666666667 20110016 D2 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 4 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-D2 3.0166666666666666 20110016 E03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 7 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E03 2.683333333333333 20110016 E38 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:51:17 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 1 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E38 2.1 20110016 E41 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 8 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E41 2.716666666666667 20110016 M03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:46:38 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 true null 1 MEDIC 10 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-M03 2.7666666666666666 20110016 RS1 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:46:57 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 RESCUE SQUAD 9 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-RS1 3.2666666666666666 20110016 T03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 5 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T03 3.2333333333333334 20110016 T13 2003235 Structure Fire 2002-01-11 2002-01-10 Other 01/11/2002 01:51:44 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 2 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T13 2.95 20110017 M13 2003236 Medical Incident 2002-01-11 2002-01-10 Other 01/11/2002 02:27:14 AM 700 Block of FOLSOM ST SF 94107 B03 01 2176 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7831757914875, -122.399603541822) 020110017-M13 6.933333333333334 20110019 M36 2003238 Medical Incident 2002-01-11 2002-01-10 Other 01/11/2002 02:28:30 AM 9TH ST/LARKIN ST SF 94103 B02 36 2337 1 1 2 true null 1 MEDIC 1 2 6 South of Market (37.7774938436091, -122.41629170277) 020110019-M36 6.116666666666666 20110020 E38 2003239 Outside Fire 2002-01-11 2002-01-10 Other 01/11/2002 02:15:33 AM 3000 Block of WEBSTER ST SF 94123 B04 16 3462 3 3 3 false null 1 ENGINE 1 4 2 Mar

In [0]:
raw_sf_fire_calls_df.printSchema()

root
 |-- Call Number: integer (nullable = true)
 |-- Unit ID: string (nullable = true)
 |-- Incident Number: integer (nullable = true)
 |-- CallType: string (nullable = true)
 |-- Call Date: date (nullable = true)
 |-- Watch Date: date (nullable = true)
 |-- Call Final Disposition: string (nullable = true)
 |-- Available DtTm: string (nullable = true)
 |-- Address: string (nullable = true)
 |-- City: string (nullable = true)
 |-- Zipcode of Incident: integer (nullable = true)
 |-- Battalion: string (nullable = true)
 |-- Station Area: string (nullable = true)
 |-- Box: string (nullable = true)
 |-- OrigPriority: string (nullable = true)
 |-- Priority: string (nullable = true)
 |-- Final Priority: integer (nullable = true)
 |-- ALS Unit: boolean (nullable = true)
 |-- Call Type Group: string (nullable = true)
 |-- NumAlarms: integer (nullable = true)
 |-- UnitType: string (nullable = true)
 |-- Unit sequence in call dispatch: integer (nullable = true)
 |-- Fire Prevention District: str

In [0]:
# Removing spaces from columns names.
# CallDate, WatchDate have a str data type need to change into date format plus AvailableDtTm need to convert into timestamp format


cleaned_sf_fire_calls_df = raw_sf_fire_calls_df.withColumnRenamed("Call Number", "CallNumber")\
                                               .withColumnRenamed("Unit ID", "UnitID")\
                                               .withColumnRenamed("Incident Number", "IncidentNumber")\
                                               .withColumnRenamed("Call Date", "CallDate").withColumn("CallDate", to_date("CallDate", "MM/dd/yyyy"))\
                                               .withColumnRenamed("Watch Date", "WatchDate").withColumn("WatchDate", to_date("WatchDate", "MM/dd/yyyy"))\
                                               .withColumnRenamed("Call Final Disposition", "CallFinalDisposition")\
                                               .withColumnRenamed("Available DtTm", "AvailableDtTm").withColumn("AvailableDtTm", to_timestamp("AvailableDtTm",                                                                     "MM/dd/yyyy hh:mm:ss a"))\
                                               .withColumnRenamed("Zipcode of Incident", "Zipcode")\
                                               .withColumnRenamed("Station Area", "StationArea")\
                                               .withColumnRenamed("Final Priority", "FinalPriority")\
                                               .withColumnRenamed("ALS Unit", "ALSUnit")\
                                               .withColumnRenamed("Call Type Group", "CallTypeGroup")\
                                               .withColumnRenamed("Unit sequence in call dispatch", "Unitsequenceincalldispatch")\
                                               .withColumnRenamed("Fire Prevention District", "FirePreventionDistrict")\
                                               .withColumnRenamed("Supervisor District", "SupervisorDistrict")

display(cleaned_sf_fire_calls_df)

CallNumber UnitID IncidentNumber CallType CallDate WatchDate CallFinalDisposition AvailableDtTm Address City Zipcode Battalion StationArea Box OrigPriority Priority FinalPriority ALSUnit CallTypeGroup NumAlarms UnitType Unitsequenceincalldispatch FirePreventionDistrict SupervisorDistrict Neighborhood Location RowID Delay 20110014 M29 2003234 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T01:58:43.000+0000 10TH ST/MARKET ST SF 94103 B02 36 2338 1 1 2 true null 1 MEDIC 1 2 6 Tenderloin (37.7765408927183, -122.417501464907) 020110014-M29 5.233333333333333 20110015 M08 2003233 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T02:10:17.000+0000 300 Block of 5TH ST SF 94107 B03 08 2243 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7792841462441, -122.402061300134) 020110015-M08 3.0833333333333335 20110016 B02 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 6 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B02 3.05 20110016 B04 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:51:54.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 3 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B04 2.316666666666667 20110016 D2 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 4 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-D2 3.0166666666666666 20110016 E03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 7 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E03 2.683333333333333 20110016 E38 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:51:17.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 1 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E38 2.1 20110016 E41 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 8 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E41 2.716666666666667 20110016 M03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:46:38.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 true null 1 MEDIC 10 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-M03 2.7666666666666666 20110016 RS1 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:46:57.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 RESCUE SQUAD 9 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-RS1 3.2666666666666666 20110016 T03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 5 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T03 3.2333333333333334 20110016 T13 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:51:44.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 2 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T13 2.95 20110017 M13 2003236 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T02:27:14.000+0000 700 Block of FOLSOM ST SF 94107 B03 01 2176 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7831757914875, -122.399603541822) 020110017-M13 6.933333333333334 20110019 M36 2003238 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T02:28:30.000+0000 9TH ST/LARKIN ST SF 94103 B02 36 2337 1 1 2 true null 1 MEDIC 1 2 6 South of Market (37.7774938436091, -122.41629170277) 020110019-M36 6.116666666666666 20110020 E38 2003239 Outside Fire 2002-01-11 2002-01-10 Other 2002-01-11T02:15:33.000+0000 3000 Block of WEBSTER S

In [0]:
cleaned_sf_fire_calls_df.printSchema()

root
 |-- CallNumber: integer (nullable = true)
 |-- UnitID: string (nullable = true)
 |-- IncidentNumber: integer (nullable = true)
 |-- CallType: string (nullable = true)
 |-- CallDate: date (nullable = true)
 |-- WatchDate: date (nullable = true)
 |-- CallFinalDisposition: string (nullable = true)
 |-- AvailableDtTm: timestamp (nullable = true)
 |-- Address: string (nullable = true)
 |-- City: string (nullable = true)
 |-- Zipcode: integer (nullable = true)
 |-- Battalion: string (nullable = true)
 |-- StationArea: string (nullable = true)
 |-- Box: string (nullable = true)
 |-- OrigPriority: string (nullable = true)
 |-- Priority: string (nullable = true)
 |-- FinalPriority: integer (nullable = true)
 |-- ALSUnit: boolean (nullable = true)
 |-- CallTypeGroup: string (nullable = true)
 |-- NumAlarms: integer (nullable = true)
 |-- UnitType: string (nullable = true)
 |-- Unitsequenceincalldispatch: integer (nullable = true)
 |-- FirePreventionDistrict: string (nullable = true)
 |-- S

#### Q1. How many distinct typs of calls were made to the fire dept ?

In [0]:
# By using countDistinct method 
q1_df = cleaned_sf_fire_calls_df.where("CallType is not null")\
                                .select(countDistinct("CallType"))

display(q1_df)

# Other method 
q1_df1 = cleaned_sf_fire_calls_df.where("CallType is not null")\
                                .select("CallType")\
                                .distinct()
print(q1_df1.count())

count(DISTINCT CallType) 32

32


#### Q2. What are the different distinct types of calls made to the fire dipt ?

In [0]:
q2_df = cleaned_sf_fire_calls_df.where("CallType is not null")\
                                .select("CallType")\
                                .distinct()

display(q2_df)

CallType Elevator / Escalator Rescue Marine Fire Aircraft Emergency Confined Space / Structure Collapse Administrative Alarms Odor (Strange / Unknown) Citizen Assist / Service Call HazMat Watercraft in Distress Explosion Oil Spill Vehicle Fire Suspicious Package Extrication / Entrapped (Machinery, Vehicle) Other Outside Fire Traffic Collision Assist Police Gas Leak (Natural and LP Gases) Water Rescue Electrical Hazard High Angle Rescue Structure Fire Industrial Accidents Medical Incident Mutual Aid / Assist Outside Agency Fuel Spill Smoke Investigation (Outside) Train / Rail Incident Train / Rail Fire Lightning Strike (Investigation)

#### Q3. Find out all the responses or delayed times greater than 5 mins

In [0]:
q3_df = cleaned_sf_fire_calls_df.select("CallNumber", "delay")\
                                .where("delay > 5")

display(q3_df)

CallNumber delay 20110014 5.233333333333333 20110017 6.933333333333334 20110019 6.116666666666666 20110039 7.85 20110045 77.33333333333333 20110046 5.416666666666667 20110055 6.5 20110058 6.85 20110058 6.85 20110061 6.333333333333333 20110062 7.016666666666667 20110066 9.1 20110068 5.983333333333333 20110070 5.416666666666667 20110070 5.083333333333333 20110070 8.183333333333334 20110070 5.983333333333333 20110077 18.066666666666666 20110103 5.55 20110103 5.966666666666667 20110103 6.05 20110141 6.133333333333334 20110148 6.266666666666667 20110157 6.3 20110162 5.633333333333334 20110163 5.05 20110163 5.633333333333334 20110170 12.733333333333333 20110180 6.15 20110180 7.3 20110190 5.383333333333334 20110196 8.0 20110196 7.766666666666667 20110198 5.25 20110237 5.05 20110240 7.516666666666667 20110245 5.916666666666667 20110252 5.2 20110268 21.116666666666667 20110271 5.216666666666667 20110273 5.4 20110276 8.7 20110277 8.216666666666667 20110277 8.45 20110289 10.3 20110297 6.5 20110310 5.033333333333333 20110315 5.35 20110334 5.1 20110336 11.066666666666666 20110338 28.25 20110338 6.45 20110338 6.3 20110338 7.25 20110338 6.25 20110342 7.0 20110348 8.65 20110349 7.316666666666666 20110357 5.066666666666666 20110360 6.55 20110367 11.383333333333333 20110367 10.533333333333333 20110367 10.566666666666666 20110368 5.666666666666667 20110369 8.366666666666667 20110378 12.516666666666667 20110383 5.316666666666666 20110388 8.55 20110389 16.716666666666665 20110390 5.35 20110400 12.7 20110405 6.0 20110410 8.283333333333333 20110421 10.883333333333333 20110421 10.566666666666666 20110424 7.7 20110424 5.066666666666666 20110424 5.2 20110427 8.633333333333333 20110429 6.65 20110435 6.066666666666666 20110437 7.533333333333333 20110455 20.366666666666667 20110469 5.516666666666667 20110473 7.333333333333333 20110475 5.183333333333334 20120026 7.266666666666667 20120027 5.3 20120031 5.55 20120034 5.916666666666667 20120038 17.016666666666666 20120039 9.1 20120040 6.033333333333333 20120047 10.2 20120059 9.116666666666667 20120059 8.583333333333334 20120076 5.65 20120104 7.266666666666667 20120108 6.566666666666666 20120137 11.716666666666667 20120147 6.25 20120153 7.45 20120155 5.333333333333333 20120156 14.666666666666666 20120156 8.95 20120163 7.05 20120164 5.516666666666667 20120173 6.65 20120173 9.6 20120173 8.35 20120173 6.95 20120191 5.666666666666667 20120195 6.5 20120206 5.166666666666667 20120208 7.4 20120213 5.633333333333334 20120230 7.566666666666666 20120233 5.05 20120235 41.36666666666667 20120247 6.133333333333334 20120260 5.683333333333334 20120268 5.183333333333334 20120272 8.166666666666666 20120284 5.066666666666666 20120296 5.066666666666666 20120312 6.2 20120334 5.3 20120339 5.266666666666667 20120342 5.033333333333333 20120349 6.233333333333333 20130013 5.2 20130020 20.566666666666666 20130041 5.416666666666667 20130041 5.733333333333333 20130042 6.166666666666667 20130042 5.416666666666667 20130047 5.583333333333333 20130053 13.816666666666666 20130054 5.55 20130055 8.0 20130063 9.283333333333333 20130065 9.2 20130068 8.433333333333334 20130077 6.716666666666667 20130094 7.466666666666667 20130109 7.316666666666666 20130122 6.35 20130131 6.25 20130136 10.1 20130140 5.033333333333333 20130153 5.583333333333333 20130159 5.65 20130165 5.233333333333333 20130170 6.316666666666666 20130178 5.616666666666666 20130180 8.683333333333334 20130182 5.316666666666666 20130183 5.3 20130187 7.766666666666667 20130195 7.433333333333334 20130198 20.983333333333334 20130202 6.1 20130216 12.233333333333333 20130219 5.3 20130220 5.216666666666667 20130224 9.1 20130225 8.35 20130230 8.433333333333334 20130247 17.216666666666665 20130260 6.2 20130260 6.3 20130268 5.833333333333333 20130268 5.083333333333333 20130272 8.133333333333333 20130274 8.35 20130274 8.35 20130274 8.35 20130290 6.483333333333333 20130291 7.966666666666667 20130293 5.95 20130295 11.7 20130297 5.383333333333334 20130301 5.6 20130301 6.3 20130318 5.216666666666667 2

#### Q4. What were the most common call types ?

In [0]:
q4_df = cleaned_sf_fire_calls_df.where("CallType is not null")\
                                .groupBy("CallType")\
                                .count()\
                                .orderBy("count", ascending=False)
                                
display(q4_df)

CallType count Medical Incident 2843475 Structure Fire 578998 Alarms 483518 Traffic Collision 175507 Citizen Assist / Service Call 65360 Other 56961 Outside Fire 51603 Vehicle Fire 20939 Water Rescue 20037 Gas Leak (Natural and LP Gases) 17284 Electrical Hazard 12608 Elevator / Escalator Rescue 11851 Odor (Strange / Unknown) 11680 Smoke Investigation (Outside) 9796 Fuel Spill 5198 HazMat 3437 Industrial Accidents 2401 Explosion 2307 Aircraft Emergency 1196 Train / Rail Incident 1116 Assist Police 1058 High Angle Rescue 1051 Watercraft in Distress 728 Extrication / Entrapped (Machinery, Vehicle) 628 Oil Spill 441 Confined Space / Structure Collapse 438 Marine Fire 331 Mutual Aid / Assist Outside Agency 298 Suspicious Package 266 Administrative 120 Train / Rail Fire 23 Lightning Strike (Investigation) 6

#### Q5. What ZIP codes accounted for the most commom calls ?

In [0]:
q5_df = cleaned_sf_fire_calls_df.where("CallType is not null")\
                                .groupBy("CallType", "Zipcode")\
                                .count()\
                                .orderBy("count", ascending=False)

display(q5_df)

CallType Zipcode count Medical Incident 94102 401457 Medical Incident 94103 370215 Medical Incident 94110 249279 Medical Incident 94109 238087 Medical Incident 94124 147564 Medical Incident 94112 139565 Medical Incident 94115 120087 Medical Incident 94122 107602 Medical Incident 94107 107439 Medical Incident 94133 99050 Medical Incident 94117 92744 Medical Incident 94134 83569 Medical Incident 94114 82378 Medical Incident 94118 77817 Medical Incident 94121 74943 Medical Incident 94116 66742 Medical Incident 94132 64439 Structure Fire 94110 57014 Medical Incident 94105 56909 Structure Fire 94103 55529 Structure Fire 94102 55134 Alarms 94102 54481 Medical Incident 94108 53566 Alarms 94103 53183 Structure Fire 94109 52240 Medical Incident 94123 48001 Medical Incident 94131 47067 Alarms 94109 43803 Medical Incident 94111 38744 Structure Fire 94124 37585 Structure Fire 94112 34359 Alarms 94115 33518 Medical Incident 94127 29159 Structure Fire 94115 26214 Structure Fire 94117 24990 Alarms 94110 24381 Alarms 94105 23589 Alarms 94108 23363 Alarms 94107 22729 Structure Fire 94107 22032 Traffic Collision 94103 21641 Structure Fire 94114 21307 Structure Fire 94133 21046 Alarms 94133 20964 Structure Fire 94122 18265 Structure Fire 94134 17750 Structure Fire 94118 17714 Traffic Collision 94110 16830 Alarms 94117 16633 Alarms 94132 16631 Alarms 94118 16551 Structure Fire 94108 16524 Structure Fire 94123 16253 Alarms 94123 16086 Medical Incident 94104 15973 Alarms 94111 15524 Structure Fire 94121 14499 Medical Incident 94130 14458 Structure Fire 94131 12812 Alarms 94114 12223 Alarms 94131 11812 Traffic Collision 94102 11760 Traffic Collision 94109 11759 Alarms 94124 11757 Structure Fire 94105 11493 Structure Fire 94132 11472 Structure Fire 94116 11448 Traffic Collision 94112 11134 Alarms 94121 10965 Medical Incident 94158 10720 Alarms 94122 9930 Alarms 94104 9572 Traffic Collision 94124 9295 Traffic Collision 94122 8971 Alarms 94112 8763 Traffic Collision 94107 8156 Structure Fire 94111 7550 Traffic Collision 94118 7147 Alarms 94134 7032 Traffic Collision 94132 6662 Outside Fire 94124 6600 Structure Fire 94127 6471 Traffic Collision 94105 6353 Alarms 94116 5943 Traffic Collision 94117 5911 Traffic Collision 94115 5609 Other 94110 5267 Traffic Collision 94130 5106 Citizen Assist / Service Call 94109 5097 Alarms 94127 4915 Citizen Assist / Service Call 94102 4893 Traffic Collision 94134 4828 Citizen Assist / Service Call 94110 4746 Citizen Assist / Service Call 94103 4674 Traffic Collision 94114 4656 Alarms 94158 4625 Medical Incident 94129 4583 Structure Fire 94104 4578 Outside Fire 94103 4560 Other 94103 4534 Traffic Collision 94133 4391 Traffic Collision 94121 4374 Other 94102 4197 Outside Fire 94110 4137 Other 94109 3969 Outside Fire 94111 3960 Traffic Collision 94116 3896 Citizen Assist / Service Call 94124 3862 Other 94124 3722 Vehicle Fire 94124 3684 Citizen Assist / Service Call 94115 3660 Citizen Assist / Service Call 94112 3634 Outside Fire 94107 3459 Outside Fire 94133 3412 Citizen Assist / Service Call 94122 3410 Traffic Collision 94123 3217 Traffic Collision 94131 3118 Citizen Assist / Service Call 94118 3103 Water Rescue 94121 3089 Citizen Assist / Service Call 94116 2866 Other 94112 2768 Other 94115 2756 Traffic Collision 94108 2706 Other 94122 2651 Citizen Assist / Service Call 94117 2545 Outside Fire 94102 2543 Citizen Assist / Service Call 94114 2539 Alarms 94129 2466 Traffic Collision 94111 2442 Citizen Assist / Service Call 94121 2440 Elevator / Escalator Rescue 94102 2321 Outside Fire 94122 2308 Water Rescue 94129 2304 Other 94107 2284 Outside Fire 94109 2251 Other 94114 2237 Citizen Assist / Service Call 94107 2232 Other 94117 2208 Citizen Assist / Service Call 94131 2199 Citizen Assist / Service Call 94132 2167 Other 94118 2164 Citizen Assist / Service Call 94133 2144 Citizen Assist / Service Call 94134 2130 Outside Fire 94134 2095 Other 94133 2075 Citizen Assist / Service Call 94123 2070 Alarms 94130 2024 Water Rescue 9

#### Q6. What San Francisco neighborhoods are in the ZIP codes 94102 and 94103 ?

In [0]:
q6_df = cleaned_sf_fire_calls_df.where("Zipcode == 94102 or Zipcode == 94102")\
                                .select("Neighborhood", "Zipcode")

display(q6_df)

Neighborhood Zipcode Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Western Addition 94102 Hayes Valley 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Financial District/South Beach 94102 Financial District/South Beach 94102 Tenderloin 94102 Financial District/South Beach 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Western Addition 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Western Addition 94102 Western Addition 94102 South of Market 94102 South of Market 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Nob Hill 94102 Nob Hill 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Western Addition 94102 Western Addition 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Tenderloin 94102 Hayes Valley 94102 Hayes Valley 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Hayes Valley 94102 Hayes Valley 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Western Addition 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Hayes Valley 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 South of Market 94102 South of Market 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Tenderloin 94102 Tenderloin 94102 Hayes Valley 94102 Hayes Valley 94102 Nob Hill 94102 Nob Hill 94102 Nob Hill 94102 Tenderloi

#### Q7. What was the sum of all call alarms, average, min and max of the call response time ?

In [0]:
q7_df = cleaned_sf_fire_calls_df.selectExpr("sum(NumAlarms) as sum_alrams", "avg(delay) as avg_delay", "min(delay) as min_delay", "max(delay) as max_delay")
                                
display(q7_df)                              

sum_alrams avg_delay min_delay max_delay 4403441 3.902170335063649 0.016666666666666666 1879.6166666666666

#### Q8. How many distinct years of data in CSV files ?

In [0]:
q8_df = cleaned_sf_fire_calls_df.withColumn("years", year("CallDate"))\
                                .select("years").distinct()

display(q8_df)

years 2003 2006 2005 2002 2009 2010 2011 2008 2013 2012 2015 2014 2016 2018 2017 2007 2004 2001 2000

#### Q9. What week of the year in 2018 had the most fire calls ?

In [0]:
q9_df = cleaned_sf_fire_calls_df.withColumn("year", year("CallDate"))\
                                .where("year== 2018")\
                                .withColumn("week_of_year", weekofyear("CallDate"))\
                                .groupBy("week_of_year")\
                                .count()\
                                .select("week_of_year", "count")\
                                .orderBy("count", ascending=False)

display(q9_df)

week_of_year count 1 6401 25 6163 13 6103 22 6060 44 6048 27 6042 16 6009 40 6000 43 5986 5 5946 2 5929 18 5917 9 5874 8 5843 6 5839 21 5821 38 5817 10 5806 23 5781 32 5764 29 5761 14 5752 28 5749 7 5739 4 5738 15 5724 37 5705 11 5690 26 5686 3 5676 24 5624 19 5619 17 5598 42 5587 12 5579 41 5551 34 5493 36 5438 35 5422 20 5390 30 5389 39 5329 31 5328 33 5314 45 1572

#### Q10. What neighborhoods in San Francisco had the worst respinse time in 2018 ?

In [0]:
q10_df = cleaned_sf_fire_calls_df.withColumn("year", year("CallDate"))\
                                 .where("year==2018")\
                                 .select("Neighborhood", "delay")\
                                 .orderBy("delay", ascending=False)

display(q10_df)

Neighborhood delay West of Twin Peaks 754.0833333333334 Mission 745.9333333333333 Chinatown 734.8666666666667 Bayview Hunters Point 715.7666666666667 Bayview Hunters Point 714.7333333333333 Bayview Hunters Point 713.05 Bayview Hunters Point 700.0 Mission 669.4166666666666 Bayview Hunters Point 609.0833333333334 Castro/Upper Market 561.1666666666666 West of Twin Peaks 539.9833333333333 Outer Richmond 508.85 Chinatown 506.95 Chinatown 491.45 Chinatown 491.26666666666665 Chinatown 490.03333333333336 Chinatown 485.8 Sunset/Parkside 457.4 Financial District/South Beach 419.73333333333335 Financial District/South Beach 419.6333333333333 Financial District/South Beach 419.6333333333333 Financial District/South Beach 419.6333333333333 Financial District/South Beach 419.6333333333333 Financial District/South Beach 419.6333333333333 Financial District/South Beach 419.6333333333333 Financial District/South Beach 419.6333333333333 Financial District/South Beach 418.5833333333333 Financial District/South Beach 417.8666666666667 Financial District/South Beach 417.71666666666664 Financial District/South Beach 415.7 Financial District/South Beach 409.31666666666666 Financial District/South Beach 409.31666666666666 Financial District/South Beach 406.96666666666664 Financial District/South Beach 406.6333333333333 Financial District/South Beach 406.31666666666666 Financial District/South Beach 405.6 Financial District/South Beach 405.1 Financial District/South Beach 404.71666666666664 Financial District/South Beach 404.1 Financial District/South Beach 403.9 Financial District/South Beach 402.4 Financial District/South Beach 401.73333333333335 Lakeshore 378.06666666666666 Castro/Upper Market 370.1 Sunset/Parkside 356.6166666666667 West of Twin Peaks 354.21666666666664 West of Twin Peaks 349.45 Inner Sunset 346.6666666666667 Inner Sunset 346.5 Sunset/Parkside 345.15 Inner Sunset 344.23333333333335 Inner Sunset 343.18333333333334 Tenderloin 340.48333333333335 Potrero Hill 336.65 Bayview Hunters Point 319.53333333333336 Outer Mission 290.3833333333333 Treasure Island 266.46666666666664 Lakeshore 253.9 Chinatown 247.51666666666668 Chinatown 247.51666666666668 Chinatown 247.51666666666668 Chinatown 247.51666666666668 Chinatown 247.51666666666668 Castro/Upper Market 239.65 Sunset/Parkside 203.43333333333334 Castro/Upper Market 202.25 Lakeshore 201.28333333333333 Pacific Heights 196.1 Russian Hill 192.95 Potrero Hill 182.35 Mission 179.41666666666666 Haight Ashbury 175.86666666666667 Inner Sunset 174.31666666666666 Inner Sunset 174.08333333333334 West of Twin Peaks 172.98333333333332 Inner Sunset 172.16666666666666 Inner Sunset 169.61666666666667 Mission 169.25 West of Twin Peaks 164.45 West of Twin Peaks 164.18333333333334 Treasure Island 160.21666666666667 Portola 157.96666666666667 Bayview Hunters Point 155.8 Bayview Hunters Point 155.03333333333333 Mission 151.98333333333332 Outer Mission 151.38333333333333 Lakeshore 151.1 Potrero Hill 144.88333333333333 Bernal Heights 143.45 Chinatown 142.81666666666666 Chinatown 141.73333333333332 Chinatown 141.73333333333332 Bayview Hunters Point 141.1 Mission 140.28333333333333 Financial District/South Beach 135.51666666666668 Outer Mission 133.78333333333333 Sunset/Parkside 133.11666666666667 South of Market 129.31666666666666 Bernal Heights 129.01666666666668 Pacific Heights 129.01666666666668 Bernal Heights 128.95 Bernal Heights 128.48333333333332 Bernal Heights 128.41666666666666 Bernal Heights 128.03333333333333 Outer Mission 124.16666666666667 Tenderloin 123.43333333333334 Pacific Heights 123.05 Pacific Heights 123.05 Financial District/South Beach 120.11666666666666 Castro/Upper Market 119.75 Financial District/South Beach 119.46666666666667 West of Twin Peaks 116.86666666666666 Mission 116.28333333333333 Outer Mission 116.1 Noe Valley 115.95 Tenderloin 115.7 South of Market 113.9 Mission Bay 113.25 Mission 113.01666666666667 Inner Richmond 112.9 South of Market 112.53333333333333 Noe Valley 112.35 Bayvie